File: C/cp/catalan_number.c
/*
The Catalan numbers form a sequence of natural numbers that occur in various 
counting problems, often involving recursively-defined objects.

Application in some combinatorial problems:
1.Number of correct bracket sequence consisting of n opening and n closing brackets.
2.The number of rooted full binary trees with n+1 leaves (vertices are not numbered). A rooted binary tree is full if every vertex has either two children or no children.
3.The number of ways to completely parenthesize n+1 factors. 
*/

#include <stdio.h>

int catalan(int num)
{
	if (num <= 1)
	{
		return 1;
	}

	int res = 0;
	for (int i = 0; i < num; i++)
		res += catalan(i) *catalan(num - i - 1);
	return res;
}

int main()
{
	int num;
	printf("Enter Number: ");
	scanf("%d", &num);
	printf("\n");
	for (int i = 1; i <= num; i++)
	{
		printf("%d ", catalan(i));
	}

	return 0;
}

/*
Sample input and output
Enter number: 7
1 2 5 14 42 132 429

time complexity: O(n^2)
space complexity: O(n)
*/ 

File: C-Sharp/math/Catalan_Number.cs
// Nth Catalan Number 
using System; 
class CatalanNumber 
{ 
    // A recursive function to find 
    // nth catalan number 
    static int catalan(int n) 
    { 
        int res = 0; 
        
        // Base case 
        if (n <= 1) 
        { 
            return 1; 
        } 
        for (int i = 0; i < n; i++) 
        {
            res += catalan(i) * catalan(n - i - 1); 
        } 
        return res; 
    } 
    
    // Main Function
    public static void Main() 
    { 
        int number;
        Console.Write("Enter the Number: ");
        number = int.Parse(Console.ReadLine());
        Console.Write("Nth Catalan numbers are: ");
        // Catalan Numbers
        for (int i = 0; i < number; i++) 
            Console.Write(catalan(i) + " "); 
    } 
} 

/*
Input:
Enter the Number: 4
Output:
Nth Catalan numbers are: 1 1 2 5

Time Complexity: O(2^n)
Space Complexity: O(1)
*/
